### PR TITLE
Tournament creation

### PR DESCRIPTION
**Pull request by EthanDelage**
This pull request introduces a new API endpoint, '/tournament' with the POST method, enabling the creation of tournaments. The endpoint accepts various parameters in the body, as indicated in the documentation. To create a tournament, the client must provide a valid access token in the Authorization header.

### CODE REVIEW

File: tournament/src/api/views.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        errors = []
+
+        if name is None:
+            errors.append('Missing name field')
+        if len(name) < settings.MIN_TOURNAMENT_NAME_LENGTH:
+            errors.append(f'Tournament name must contain at least {settings.MIN_TOURNAMENT_NAME_LENGTH} characters')
+        elif len(name) > settings.MAX_TOURNAMENT_NAME_LENGTH:
+            errors.append(f'Tournament name must contain less than {settings.MAX_TOURNAMENT_NAME_LENGTH} characters')
+        if not name.replace(' ', '').isalnum():
+            errors.append('Tournament name may only contain letters, numbers and spaces')
```
Comment (tdameros) [inline tournament/src/api/views.py:85]: Maybe you can use slugify tool with django

https://docs.djangoproject.com/en/5.0/ref/utils/#django.utils.text.slugify

Comment (tdameros) [inline tournament/src/api/views.py:85]: For every utils functions like staticmethod you can use type hint for more clarity and assistance from your IDE.

Example :


```
from typing import Optional, Tuple, Any

def is_valid_max_players(max_players: Any) -> Tuple[bool, Optional[str]]:
```

https://docs.python.org/3/library/typing.html

File: tournament/src/api/views.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                tournament.max_players = max_players
+            registration_deadline = json_request.get('registration-deadline')
+            if json_request.get('registration-deadline') is not None:
+                registration_deadline = parser.isoparse(registration_deadline)
+                tournament.registration_deadline = TournamentView.convert_to_utc_datetime(registration_deadline)
+            tournament.save()
+
+            return JsonResponse(model_to_dict(tournament), status=201)
+        except json.JSONDecodeError:
+            return JsonResponse(data={'errors': ['Invalid JSON format in request body']}, status=400)
```
Comment (tdameros) [inline tournament/src/api/views.py:46]: Why do everything in the try-except block? Personally, that's something I dislike about try-except; it feels like we end up doing everything inside it. I know it's debatable, but what do you think about this alternative?


```
def post(request):
    user, authenticate_errors = authenticate_request(request)
    if user is None:
        return JsonResponse(data={'errors': authenticate_errors},
                            status=401)
    try:
        json_request = json.loads(request.body.decode('utf8'))
    except json.JSONDecodeError:
        return JsonResponse(
            data={'errors': ['Invalid JSON format in request body']},
            status=400)
    valid_tournament, errors = TournamentView.is_valid_tournament(
        json_request)
    if not valid_tournament:
        return JsonResponse(data={'errors': errors}, status=400)

    tournament = Tournament(
        name=json_request['name'],
        is_private=json_request['is-private'],
        admin_id=user['id']
    )
    max_players = json_request.get('max-players')
    if max_players is not None:
        tournament.max_players = max_players
    registration_deadline = json_request.get('registration-deadline')
    if json_request.get('registration-deadline') is not None:
        registration_deadline = parser.isoparse(registration_deadline)
        tournament.registration_deadline = TournamentView.convert_to_utc_datetime(
            registration_deadline)
    tournament.save()
    return JsonResponse(model_to_dict(tournament), status=201)

```

**Review by tdameros — CHANGES_REQUESTED** [review]

File: tournament/src/api/views.py (hunk 1 above)
Comment (EthanDelage) [reply tournament/src/api/views.py:85]: This condition is used to check that there are only alphanumeric characters and spaces, whereas slugify() is used to modify this string, not to check its content.

File: tournament/src/api/views.py (hunk 2 above)
Comment (EthanDelage) [reply tournament/src/api/views.py:46]: that's fine for me. I didn't know that json_request remained available outside the scope of the try block.

File: tournament/src/tournament/authenticate_request.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        return decoded_payload
+    except Exception as e:
+        errors.append(str(e))
+        return None
+
+
+def get_user(user_id, encoded_jwt, errors):
+    headers = {'Authorization': encoded_jwt}
+    try:
+        response = requests.get(f'http://user-management-nginx/user/{user_id}', headers=headers)
```
Comment (aLeuleu) [inline tournament/src/tournament/authenticate_request.py]: Note to myself :  I need to implement this quickly.

File: tournament/src/tournament/authenticate_request.py
Comment (aLeuleu) [inline tournament/src/tournament/authenticate_request.py:1]: I feel like I better share my JWT manager with you guys ( maybe with a library that follows update ? ..)
You don't need to implement things that are already coded in my part.
We should all share the same JWT manager ( I'll be the only one with the secret key so thats ok)

File: tournament/tournament-api-documentation.md [hunk 4]
```diff
@@ -6,17 +6,61 @@
...
  <summary><code>POST</code> <code><b>/tournament</b></code></summary>
 
-#### Create a new tournament
+
+### Parameters
+
+#### Body
+
+- Tournament name must be between 3 and 20 characters and can only contain alnum and space
+- Max players must be between 2 and 16 (optional, default = 16 players)
```
Comment (aLeuleu) [inline tournament/tournament-api-documentation.md]: The word 'max' should be removed

**Review by aLeuleu — CHANGES_REQUESTED** [review]
Small enhancement could be made (nothing serious)
There are no unit test !
Shouldn't we request some ? (open for debate on this point)

Apart from that : clean work, good job

note to myself : I shall quickly implement ` GET /user/{user_id}/ `

File: tournament/src/api/views.py (hunk 2 above)
Comment (EthanDelage) [reply tournament/src/api/views.py]: 87c7dc3c4536ca5c9b84a5a8aeb10a5f2c05b071

File: tournament/src/api/models.py [hunk 5]
```diff
@@ -1,8 +1,14 @@
 from django.db import models
+from enum import Enum
 
 
 class Tournament(models.Model):
+    CREATED = 0
+    IN_PROGRESS = 1
+    FINISHED = 2
```
[the remark this answers is not in the document]
Comment (EthanDelage) [reply tournament/src/api/models.py:8]: I tried making class enums, but it didn't work.

File: tournament/src/tournament/authenticate_request.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    user = get_user(user_id, encoded_jwt, errors)
+    if user is not None:
+        return user, None
+    return None, errors
+
+
+def decode_jwt(encoded_jwt, errors):
+    try:
+        decoded_payload = jwt.decode(encoded_jwt, 'django-insecure-&r*!icx1$(sv7f-sj&ezvjxw+pljt-yz(r6yowfg18ihdu@15k',
+                                     algorithms=["HS256"])
```
[the remark this answers is not in the document]
Comment (EthanDelage) [reply tournament/src/tournament/authenticate_request.py]: 67d148d102712042278a9d096d7ad0854cabe7aa

File: tournament/src/tournament/authenticate_request.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        return None, errors
+    user = get_user(user_id, encoded_jwt, errors)
+    if user is not None:
+        return user, None
+    return None, errors
+
+
+def decode_jwt(encoded_jwt, errors):
+    try:
+        decoded_payload = jwt.decode(encoded_jwt, 'django-insecure-&r*!icx1$(sv7f-sj&ezvjxw+pljt-yz(r6yowfg18ihdu@15k',
```
[the remark this answers is not in the document]
Comment (EthanDelage) [reply tournament/src/tournament/authenticate_request.py]: 67d148d102712042278a9d096d7ad0854cabe7aa

File: tournament/tournament-api-documentation.md (hunk 4 above)
Comment (EthanDelage) [reply tournament/tournament-api-documentation.md]: 67d148d102712042278a9d096d7ad0854cabe7aa

File: tournament/src/api/views.py (hunk 1 above)
Comment (EthanDelage) [reply tournament/src/api/views.py]: f57f4cecafd7a8edcb636558019d1d8ea479760b

File: tournament/src/api/views.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    @staticmethod
+    def is_valid_max_players(max_players: Any) -> tuple[bool, str | None]:
+        if max_players is None:
+            return True, None
+        if not isinstance(max_players, int):
+            return False, 'Max players must be an integer'
+        if max_players > settings.MAX_PLAYERS:
+            return False, f'Max players must contain less than {settings.MAX_PLAYERS} slots'
+        if max_players < settings.MIN_PLAYERS:
+            return False, f'Max players must contain at least {settings.MIN_PLAYERS} slots'
```
Comment (hferraud) [inline tournament/src/api/views.py:101]: Error message is incorrect you should return either:
`Tournament must contain less than {settings.MAX_PLAYERS} slots`
or:
`Max players must be less than {settings.MAX_PLAYERS}`

**Review by hferraud — APPROVED** [review]

File: tournament/src/api/views.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            tournament.max_players = max_players
+        registration_deadline = json_request.get('registration-deadline')
+        if json_request.get('registration-deadline') is not None:
+            registration_deadline = parser.isoparse(registration_deadline)
+            tournament.registration_deadline = TournamentView.convert_to_utc_datetime(registration_deadline)
+        tournament.save()
+        return JsonResponse(model_to_dict(tournament), status=201)
+
+    @staticmethod
+    def is_valid_tournament(json_request: Dict[str, Any]) -> Tuple[bool, list[str]] | Tuple[bool, None]:
```
Comment (tdameros) [inline tournament/src/api/views.py:50]: `def is_valid_tournament(json_request: dict[str, Any]) -> tuple[bool, Optional[list[str]]]:
`

File: tournament/src/api/views.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            errors.append(deadline_error)
+        if not valid_private:
+            errors.append(is_private_error)
+
+        if errors:
+            return False, errors
+        return True, None
+
+    @staticmethod
+    def is_valid_name(name: Any) -> tuple[bool, list[str]] | tuple[bool, None]:
```
Comment (tdameros) [inline tournament/src/api/views.py:76]: def is_valid_name(name: Any) -> tuple[bool, Optional[list[str]]]:

File: tournament/src/api/views.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+            errors.append(f'Tournament name must contain less than {settings.MAX_TOURNAMENT_NAME_LENGTH} characters')
+        if not name.replace(' ', '').isalnum():
+            errors.append('Tournament name may only contain letters, numbers and spaces')
+
+        if errors:
+            return False, errors
+        return True, None
+
+    @staticmethod
+    def is_valid_max_players(max_players: Any) -> tuple[bool, str | None]:
```
Comment (tdameros) [inline tournament/src/api/views.py:93]: def is_valid_max_players(max_players: Any) -> tuple[bool, Optional[str]]:

File: tournament/src/api/views.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        if not isinstance(max_players, int):
+            return False, 'Max players must be an integer'
+        if max_players > settings.MAX_PLAYERS:
+            return False, f'Max players must contain less than {settings.MAX_PLAYERS} slots'
+        if max_players < settings.MIN_PLAYERS:
+            return False, f'Max players must contain at least {settings.MIN_PLAYERS} slots'
+        return True, None
+
+    @staticmethod
+    def is_valid_deadline(registration_deadline: str) -> tuple[bool, str | None]:
```
Comment (tdameros) [inline tournament/src/api/views.py:105]: def is_valid_deadline(registration_deadline: str) -> tuple[bool, Optional[str]]:

File: tournament/src/api/views.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+        except ValueError:
+            return False, 'Registration deadline not in ISO 8601 date and time format'
+
+        deadline_time = TournamentView.convert_to_utc_datetime(deadline_time)
+        if deadline_time < datetime.now(timezone.utc):
+            return False, 'Registration deadline has passed'
+        return True, None
+
+    @staticmethod
+    def is_valid_private(is_private: Any) -> tuple[bool, str | None]:
```
Comment (tdameros) [inline tournament/src/api/views.py:120]: def is_valid_private(is_private: Any) -> tuple[bool, Optional[str]]:

**Review by tdameros — CHANGES_REQUESTED** [review]

**Review by tdameros — CHANGES_REQUESTED** [review]
Use python type for default type, not typing Type. For example : tuple not Tuple

**Review by tdameros — APPROVED** [review]

**Review by aLeuleu — APPROVED** [review]
looks good to me

**Pull request merged**